mantidproject/mantid
# Implement incident spectrum fitting algorithm

**Description of work.**
This PR creates a new algorithm named FitIncidentSpectrum that will fit a curve to to a spectrum using one of four methods: GaussConvCubicSpline, CubicSpline, CubicSplineViaMantid, HowellsFunction. 
There are also some small formatting fixes that I noticed while looking in other methods

**To test:**
* simple testing, this script should produce a set of plots with four fits each using the four methods available in the new algorithm:
```
import numpy as np
import matplotlib.pyplot as plt
from mantid.simpleapi import \
    AnalysisDataService, \
    CalculateEfficiencyCorrection, \
    ConvertToPointData, \
    CreateWorkspace, \
    Divide, \
    FitIncidentSpectrum, \
    Rebin

# Create the workspace to hold the already corrected incident spectrum
incident_wksp_name = 'incident_spectrum_wksp'
binning = "%s,%s,%s" % (0.2, 0.01, 4.0)
incident_wksp = CreateWorkspace(
    OutputWorkspace=incident_wksp_name,
    NSpec=1,
    DataX=[0],
    DataY=[0],
    UnitX='Wavelength',
    VerticalAxisUnit='Text',
    VerticalAxisValues='IncidentSpectrum')
incident_wksp = Rebin(InputWorkspace=incident_wksp, Params=binning)
incident_wksp = ConvertToPointData(InputWorkspace=incident_wksp)

# Spectrum function given in Milder et al. Eq (5)
def incidentSpectrum(wavelengths, phiMax, phiEpi, alpha, lambda1, lambda2, lamdaT):
    deltaTerm = 1. / (1. + np.exp((wavelengths - lambda1) / lambda2))
    term1 = phiMax * (lambdaT**4. / wavelengths**5.) * np.exp(-(lambdaT / wavelengths)**2.)
    term2 = phiEpi * deltaTerm / (wavelengths**(1 + 2 * alpha))
    return term1 + term2

# Variables for polyethlyene moderator at 300K
phiMax = 6324
phiEpi = 786
alpha = 0.099
lambda1 = 0.67143
lambda2 = 0.06075
lambdaT = 1.58

# Add the incident spectrum to the workspace
corrected_spectrum = incidentSpectrum(incident_wksp.readX(0), phiMax, phiEpi, alpha, lambda1, lambda2, lambdaT)
incident_wksp.setY(0, corrected_spectrum)

# Calculate the efficiency correction for Alpha=0.693
# and back calculate measured spectrum
eff_wksp = CalculateEfficiencyCorrection(InputWorkspace=incident_wksp, Alpha=0.693)
measured_wksp = Divide(LHSWorkspace=incident_wksp, RHSWorkspace=eff_wksp)

# Output Plots
fig, ax = plt.subplots(2, subplot_kw={'projection': 'mantid'}, sharex=True)
ax[0].plot(measured_wksp, label="Measured Incident Spectrum")
ax[0].plot(eff_wksp, label="Effiency Correction")
ax[0].plot(incident_wksp, label="After effiency correction")
ax[0].legend()

# Fit incident spectrum
prefix = "incident_spectrum_fit_with_"

fit_gauss_conv_spline = prefix + "_gauss_conv_spline"
FitIncidentSpectrum(
    InputWorkspace=incident_wksp,
    OutputWorkspace=fit_gauss_conv_spline,
    BinningForCalc=binning,
    BinningForFit=binning,
    FitSpectrumWith="GaussConvCubicSpline")

fit_spline = prefix + "_spline"
FitIncidentSpectrum(
    InputWorkspace=incident_wksp,
    OutputWorkspace=fit_spline,
    BinningForCalc=binning,
    BinningForFit=binning,
    FitSpectrumWith="CubicSpline")

fit_mantid_spline = prefix + "_mantid_spline"
FitIncidentSpectrum(
    InputWorkspace=incident_wksp,
    OutputWorkspace=fit_mantid_spline,
    BinningForCalc=binning,
    BinningForFit=binning,
    FitSpectrumWith="CubicSplineViaMantid")

fit_howells = prefix + "_howells"
howells_wksp = FitIncidentSpectrum(
    InputWorkspace=incident_wksp,
    OutputWorkspace=fit_howells,
    BinningForCalc=binning,
    BinningForFit="0.4,0.01,2.0",
    FitSpectrumWith="HowellsFunction")

# Retrieve workspaces
wksp_fit_gauss_conv_spline = AnalysisDataService.retrieve(fit_gauss_conv_spline)
wksp_fit_spline = AnalysisDataService.retrieve(fit_spline)
wksp_fit_mantid_spline = AnalysisDataService.retrieve(fit_mantid_spline)
wksp_fit_howells = AnalysisDataService.retrieve(fit_howells)

ms = 1
ax[1].plot(incident_wksp, lw=4, label="Target spectrum")
ax[1].plot(
    wksp_fit_gauss_conv_spline,
    'o',
    ms=ms,
    wkspIndex=0,
    label="Fit: GaussConvCubicSpline")
ax[1].plot(wksp_fit_spline, 'x', wkspIndex=0, ms=ms, label="Fit: CubicSpline")
ax[1].plot(
    wksp_fit_mantid_spline,
    '+',
    wkspIndex=0,
    ms=ms,
    label="Fit: CubicSplineViaMantid")
ax[1].plot(
    wksp_fit_howells, '+', wkspIndex=0, ms=ms, label="Fit: Howells Function")
ax[1].legend()

# Show plot
plt.show()
```
* Try executing algorithm from the algorithm interface.
* Test that validation of bin ranges is prevents  inputting ranges outside of the incident spectrum range.

Fixes #26454 



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
